gateway/ln-gateway/src/lightning/lnd.rs

```diff
@@ -417,124 +422,219 @@ impl ILnRpcClient for GatewayLndClient {
+            }
+        })?;
+        let cltv_limit =
+            max_delay
+                .try_into()
+                .map_err(|error| LightningRpcError::FailedPayment {
+                    failure_reason: format!("max delay exceeds valid LND range {error:?}"),
+                })?;
+
+        let dest_features = wire_features_to_lnd_feature_vec(&invoice.destination_features)
+            .map_err(|e| LightningRpcError::FailedPayment {
+                failure_reason: e.to_string(),
             })?;
-            let final_cltv_delta = invoice.min_final_cltv_delta.try_into().map_err(|error| {
+
+        debug!("LND payment does not exist for invoice {invoice:?}, will attempt to pay");
+
+        let routes = client
+            .lightning()
+            .query_routes(QueryRoutesRequest {
+                pub_key: invoice.destination.to_string(),
+                amt_msat: send_amt_msat,
+                final_cltv_delta,
+                fee_limit: Some(FeeLimit {
+                    limit: Some(Limit::FixedMsat(fee_limit_msat)),
+                }),
+                cltv_limit,
+                route_hints: route_hints_to_lnd(&invoice.route_hints),
+                dest_features,
+                use_mission_control: true,
+                ..Default::default()
+            })
+            .await
+            .map_err(|status| {
+                error!("LND failed to query routes for invoice {invoice:?} with {status:?}");
                 LightningRpcError::FailedPayment {
-                    failure_reason: format!("final cltv delta exceeds valid LND range {error:?}"),
+                    failure_reason: format!("Failed to make query routes {status:?}"),
                 }
-            })?;
-            let cltv_limit =
-                max_delay
-                    .try_into()
-                    .map_err(|error| LightningRpcError::FailedPayment {
-                        failure_reason: format!("max delay exceeds valid LND range {error:?}"),
-                    })?;
-
-            let dest_features = wire_features_to_lnd_feature_vec(&invoice.destination_features)
-                .map_err(|e| LightningRpcError::FailedPayment {
-                    failure_reason: e.to_string(),
+            })?
+            .into_inner();
+
+        let total_amt_msat: i64 =
+            invoice
+                .amount
+                .msats
+                .try_into()
+                .map_err(|error| LightningRpcError::FailedPayment {
+                    failure_reason: format!("amount exceeds valid LND amount ranges {error:?}"),
                 })?;
 
-            debug!("LND payment does not exist for invoice {invoice:?}, will attempt to pay");
-            let payments = client
+        let mut last_failed_reason: Option<String> = None;
+
+        let start = duration_since_epoch();
+        let mut attempts = 0;
+
+        for mut route in routes.routes {
```

Comment:
Do we know how this approach performs vs. the old payment code? I'd imagine iterating over a static set of routes without being able to take into account failures that happened between fetching the routes and trying any particular one would be a disadvantage over sendpay.

Generally I'd like to be very cautious here since this code is very central to the payment UX. We should likely try to get review from someone very familiar with LND's API (once we are happy with the code).

Reply:
I believe it is functionally equivalent but not 100% sure. We could just use this when we aren't sending the full amount and use the old code for normal payments